File: test/integ/ClinitSideEffect.cpp
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "AnnoUtils.h"
#include "MethodUtil.h"
#include "RedexTest.h"
#include "Show.h"

class ClinitSideEffectTest : public RedexIntegrationTest {};

TEST_F(ClinitSideEffectTest, test) {
  auto scope = build_class_scope(stores);
  auto annotation = DexType::get_type("Lcom/facebook/redextest/HasSideffects;");
  for (auto cls : scope) {
    auto has_side_effect = method::clinit_may_have_side_effects(cls);
    EXPECT_EQ(has_side_effect, get_annotation(cls, annotation) != nullptr)
        << show(cls) << " has_side_effect = " << has_side_effect;
  }
}

File: test/integ/ClinitSideEffect.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.redextest;

import java.lang.annotation.Target;
import java.lang.annotation.ElementType;

@Target(ElementType.TYPE)
@interface HasSideffects {
}

class ClinitOnlyTouchOwnFields {
  static String s_str = "";
  static int s_a = 0;
}

class ExtendNoSideEffect extends ClinitOnlyTouchOwnFields {
  static double s_d = 0.0;
}

@HasSideffects
class ClinitHasInvoke {
  static {
    System.out.println("ClinitHasInvoke.<clinit>");
  }
}

@HasSideffects
class ExtendSideEffect extends ClinitHasInvoke {
}

// Can improve
@HasSideffects
class ClinitTouchOtherFields {
  static String s_str = ClinitOnlyTouchOwnFields.s_str;
}
